utils/http-utils/src/main/scala/pl/touk/nussknacker/http/backend/HttpClientConfig.scala

```diff
@@ -75,18 +82,49 @@ object HttpClientConfig {
       useNative = forOption[Boolean]("useNative"),
       followRedirect = forOption[Boolean]("followRedirect"),
       forceShutdown = forOption[Boolean]("forceShutdown"),
-      configForProcess = forOption("configForProcess")(mapValueReader(vr))
+      configForProcess = forOption("configForProcess")(mapValueReader(vr)),
+      forbiddenHosts = forOption[List[String]]("forbiddenHosts"),
     )
   })
 
 }
 
 object DefaultHttpClientConfig {
 
-  def apply(): HttpClientConfig = HttpClientConfig(None, None, None, None, None, None, None)
+  def apply(): HttpClientConfig = HttpClientConfig(None, None, None, None, None, None, None, None)
 
   val maxPoolSize: Int = 20
 
   val timeout: FiniteDuration = 10 seconds
 
+  implicit class ClientConfigBuilderExtension(val builder: DefaultAsyncHttpClientConfig.Builder) extends AnyVal {
+
+    def setForbiddenHostRequestFilter(forbiddenHosts: List[String]): DefaultAsyncHttpClientConfig.Builder =
+      if (forbiddenHosts.nonEmpty) {
+        addForbiddenHostsFilter(forbiddenHosts)
+      } else {
+        builder
+      }
+
+    private def addForbiddenHostsFilter(forbiddenHosts: List[String]): DefaultAsyncHttpClientConfig.Builder = {
+      val lowerCaseHosts = forbiddenHosts.map(_.toLowerCase)
+      val lowCaseHostsWithMaybeLoopback = if (lowerCaseHosts.contains("localhost")) {
+        InetAddress.getLoopbackAddress.getHostAddress :: lowerCaseHosts
```

Comment:
I don't know why you need this feature, so I'll just leave this here: this will return a single localhost address, so if you are using "localhost" to lock down access for an application it will fail in multiple ways:

* a host with dual network stack (IPv4 + IPv6) has at least two localhost addresses
* on IPv4 network all 127/8 addresses point to localhost
* on modern linux with systemd-resolved you have additional domains that point to local IP: `*.localhost` and `.localhost.localdomain`
* an external domain may point to localhost - e.g. `127.0.0.1.nip.io`

A correct security filter would need to operate on ranges of resolved IP addresses, not hostnames used in URLs.

If this is not a security feature then feel free to close this thread.

Reply:
It's also security feature and feature for cloud. I changed the implementation and I think it's better than before

Reply:
Ok, just making sure:
* you know all forbidden addresses
* when you add a new internal service you will remember to extend this list
* all forbidden non-localhost services actually require a correct `Host` header, so an external DNS record pointing to your internal service will never be a problem?
? ;)

(close this thread after reading)

Reply:
* a single hostname can map to multiple IP addresses, you are checking only one
* you depend on `networkaddress.cache.ttl` being set to `-1`, i.e. on Java never refreshing it's resolved IPs (you always cache some results, but `AsyncHttpClient` will do whatever Java is configured to do), and this limitation is not documented

Reply:
Wouldn't it be better to add restrictions based on IP address that HTTP Client is connecting to? That way you'd be able to have a single forbidden IP address list with all possible private address ranges, and only one integration point in HTTP client? Address list could be a list of IP address regexes (simple but difficult to use) or proper CIDR masks.

Reply:
(seeing as you already have approval, feel free to close this thread if you want)

Reply:
Sorry, I should've just resolved my threads.

Reply:
Oh, only I meant to convey that I'm not part of product team, so I don't want my thread to block anybody. I don't know whether this PR is a band-aid or is meant to correctly lock out cloud users from accessing internal network.